src/util/address.rs

```diff
@@ -165,23 +341,17 @@ impl Payload {
     /// Get a [Payload] from an output script (scriptPubkey).
     pub fn from_script(script: &script::Script) -> Option<Payload> {
         Some(if script.is_p2pkh() {
-            Payload::PubkeyHash(PubkeyHash::from_slice(&script.as_bytes()[3..23]).unwrap())
+            let mut hash_inner = [0u8; 20];
+            hash_inner.copy_from_slice(&script.as_bytes()[3..23]);
+            Payload::PubkeyHash(PubkeyHash::from_inner(hash_inner))
         } else if script.is_p2sh() {
-            Payload::ScriptHash(ScriptHash::from_slice(&script.as_bytes()[2..22]).unwrap())
+            let mut hash_inner = [0u8; 20];
+            hash_inner.copy_from_slice(&script.as_bytes()[2..22]);
+            Payload::ScriptHash(ScriptHash::from_inner(hash_inner))
         } else if script.is_witness_program() {
-            // We can unwrap the u5 check and assume script length
-            // because [Script::is_witness_program] makes sure of this.
             Payload::WitnessProgram {
-                version: {
-                    // Since we passed the [is_witness_program] check,
-                    // the first byte is either 0x00 or 0x50 + version.
-                    let mut verop = script.as_bytes()[0];
-                    if verop > 0x50 {
-                        verop -= 0x50;
-                    }
-                    bech32::u5::try_from_u8(verop).expect("checked before")
-                },
-                program: script.as_bytes()[2..].to_vec(),
+                version: WitnessVersion::from_opcode(opcodes::All::from(script[0])).ok()?,
```

Comment:
We should avoid opcode roundtrip with a simple `from_num`

> version: WitnessVersion::from_num(script[0]).ok()?,

Reply:
We can't since it is not a version number but an instruction byte (i.e. version 1 is encoded as `0x50`).

Reply:
Off-topic, just some food for thought for the future:

This code bothered me as well but couldn't point a finger on it before. The fact that this is a question strengthens it. It seems to me that `script` has more foot-gunny API than what would be ideal. I guess it'd be safer to model it similarly to `str` which doesn't implement indexing operator because it's unclear what it should return - `char` or `u8`.  Similarly I guess `script` should have explicit `.bytes()` and `.opcodes()` iterators. It'd still not prevent someone from using it incorrectly but at least correct usage would be less clunky. Not sure if there's a better way to make it safer.

Reply:
Agreed, I got hit by this footgun :)